chainstore.go

```diff
@@ -1,147 +1,231 @@
 package chainstore
 
 import (
-	"errors"
-	"io/ioutil"
+	"fmt"
 	"regexp"
+	"sync"
+
+	"golang.org/x/net/context"
 )
 
-var (
-	ErrInvalidKey = errors.New("Invalid key")
+type storeFn func(s Store) error
 
-	KeyInvalidator = regexp.MustCompile(`(i?)[^a-z0-9\/_\-:\.]`)
+var (
+	keyInvalidator = regexp.MustCompile(`(i?)[^a-z0-9\/_\-:\.]`)
 )
 
 const (
-	MaxKeyLen = 256
+	maxKeyLen = 256
 )
 
 type Store interface {
 	Open() error
 	Close() error
-	Put(key string, val []byte) error
-	Get(key string) ([]byte, error)
-	Del(key string) error
+	Put(ctx context.Context, key string, val []byte) error
+	Get(ctx context.Context, key string) ([]byte, error)
+	Del(ctx context.Context, key string) error
 }
 
-// TODO: how can we check if a store has been opened...?
+type storeWrapper struct {
+	Store
+	errE  error
+	errMu sync.RWMutex
+}
 
+func (s *storeWrapper) err() error {
+	s.errMu.RLock()
+	defer s.errMu.RUnlock()
+	return s.errE
+}
+
+func (s *storeWrapper) setErr(err error) {
+	if err == nil {
+		return
+	}
+	s.errMu.Lock()
+	defer s.errMu.Unlock()
+	s.errE = err
+}
+
+// Chain represents a store chain.
 type Chain struct {
-	stores []Store
-	async  bool
+	stores []*storeWrapper
 }
 
+// New creates a new store chain backed by the passed stores.
 func New(stores ...Store) Store {
-	return &Chain{stores, false}
-	// TODO: make the chain..
-	// call Open(), but in case of error..?
+	c := &Chain{
+		stores: make([]*storeWrapper, 0, len(stores)),
+	}
+	for _, s := range stores {
+		c.stores = append(c.stores, &storeWrapper{Store: s})
+	}
+	return c
 }
 
-func Async(stores ...Store) Store {
-	return &Chain{stores, true}
-}
+// Open opens all the stores.
+func (c *Chain) Open() error {
 
-func (c *Chain) Open() (err error) {
-	for _, s := range c.stores {
-		err = s.Open()
-		if err != nil {
-			return // return first error that comes up
-		}
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	return
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, c *Chain, i int) {
+			s := c.stores[i]
+
+			s.setErr(s.Open())
+
+			wg.Done()
```

Comment:
put `wg.Done()` in defer function on top of goroutine --- so it's called even in case of `panic()`